### PR TITLE
[SPARK-24327][SQL] Verify and normalize a partition column name based on the JDBC resolved schema

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr modified JDBC datasource code to verify and normalize a partition column based on the JDBC resolved schema before building `JDBCRelation`.

Closes #20370

## How was this patch tested?
Added tests in `JDBCSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
@gatorsmile @conorbmurphy

**Comment by maropu** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 1]
```diff
@@ -78,7 +79,12 @@ private[sql] object JDBCRelation extends Logging {
     // Overflow and silliness can happen if you subtract then divide.
     // Here we get a little roundoff, but that's (hopefully) OK.
     val stride: Long = upperBound / numPartitions - lowerBound / numPartitions
-    val column = partitioning.column
+    val column = if (jdbcOptions.quotePartitionColumnName) {
+      val dialect = JdbcDialects.get(jdbcOptions.url)
+      dialect.quoteIdentifier(partitioning.column)
```
[the remark this answers is not in the document]
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: You point out the case users explicitly add quotes in `partitionColumn`? e.g.,
```
+      val df = spark.read.format("jdbc")
+        ...
+        .option("partitionColumn", """"THEID"""")
+        ...
+        .option("quotePartitionColumnName", quotePartitionColumnName)
+        .load()
```

[the remark this answers is not in the document]
Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: Yeah

[the remark this answers is not in the document]
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: ok, I will

[the remark this answers is not in the document]
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: [The latest fix](https://github.com/apache/spark/pull/21379/commits/8d5fa9a25dff75327e8ff8ff13b756b895e6e512) changes an existing behaviour (when quoting non-partition column names), so I'm not sure this fix is acceptable. Any suggestion?

[the remark this answers is not in the document]
Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: If possible, please do not break the existing behavior.  Both should work without specifying the extra option:
```
val df = spark.read.format("jdbc")
  .option("url", urlWithUserAndPass)
  .option("dbtable", "TEST.PEOPLE")
  .option("partitionColumn", "nonQuotedPrtColName")
  .option("lowerBound", 1)
  .option("upperBound", 4)
  .option("numPartitions", 3)
  .load()
```
```
val df = spark.read.format("jdbc")
  .option("url", urlWithUserAndPass)
  .option("dbtable", "TEST.PEOPLE")
  .option("partitionColumn", """"nonQuotedPrtColName"""")
  .option("lowerBound", 1)
  .option("upperBound", 4)
  .option("numPartitions", 3)
  .load()
```

Is that possible?

**Comment by gatorsmile** [comment]
Sorry, I just realized this is a wrong direction. Instead of trusting the user inputs, we should verify the user-specified partition columns by using the already fetched table schema info `val tableSchema = JDBCRDD.resolveTable(jdbcOptions)` when building `JDBCRelation`

**Comment by maropu** [comment]
You mean we need verification code there for checking if the fetched schema has a user-given partition column? If the schema does not have the column, throws `AnalysisException` or something?

**Comment by gatorsmile** [comment]
Yeah, we should do it.

**Comment by maropu** [comment]
ok, I will re-check.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 2]
```diff
@@ -1206,4 +1211,47 @@ class JDBCSuite extends SparkFunSuite
...
+    def testIncorrectJdbcPartitionColumn(partColName: String): Unit = {
+      val errMsg = intercept[AnalysisException] {
+        testJdbcParitionColumn(partColName, "THEID")
+      }.getMessage
+      assert(errMsg.contains(s"User-defined partition column $partColName not found " +
+        "in the JDBC relation:"))
+    }
+
+    testIncorrectJdbcPartitionColumn("NoExistingColumn")
+    withSQLConf("spark.sql.caseSensitive" -> "true") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: `SQLConf.CASE_SENSITIVE.key`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 3]
```diff
@@ -99,10 +115,21 @@ private[sql] object JDBCRelation extends Logging {
     }
     ans.toArray
   }
+
+  def getSchema(jdbcOptions: JDBCOptions, resolver: Resolver): StructType = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: Add the function description?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 4]
```diff
@@ -51,7 +53,11 @@ private[sql] object JDBCRelation extends Logging {
    * @param partitioning partition information to generate the where clause for each partition
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: Add the other two `@param` for the new parameters?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 5]
```diff
@@ -78,7 +84,17 @@ private[sql] object JDBCRelation extends Logging {
...
+
+    // Verify and normalize a partition column based on the JDBC resolved schema
+    val dialect = JdbcDialects.get(jdbcOptions.url)
+    val column = schema.map(_.name).find { fieldName =>
+      resolver(fieldName, partitioning.column) ||
+        resolver(dialect.quoteIdentifier(fieldName), partitioning.column)
+    }.map(dialect.quoteIdentifier).getOrElse {
+      throw new AnalysisException(s"User-defined partition column ${partitioning.column} not " +
+        s"found in the JDBC relation: ${schema.simpleString(Utils.maxNumToStringFields)}")
+    }
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: Create a new private function for the above resolution and checking logic?

**Comment by gatorsmile** [comment]
@maropu The fix looks good to me. Thanks for changing the solution. Could you update the PR title and description?

**Comment by maropu** [comment]
oh, I forgot to update the title..yea, I'll do soon.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala (hunk 4 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: ok

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 6]
```diff
@@ -309,7 +309,8 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
     val parts: Array[Partition] = predicates.zipWithIndex.map { case (part, i) =>
       JDBCPartition(part, i) : Partition
     }
-    val relation = JDBCRelation(parts, options)(sparkSession)
+    val schema = JDBCRelation.getSchema(sparkSession.sessionState.conf.resolver, options)
+    val relation = JDBCRelation(schema, parts, options)(sparkSession)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: We do not need to change this. Add an apply function to object JDBCRelation.scala
```
  def apply(parts: Array[Partition], jdbcOptions: JDBCOptions)(
      sparkSession: SparkSession): JDBCRelation = {
    val schema = getSchema(jdbcOptions, sparkSession.sessionState.conf.resolver)
    JDBCRelation(schema, parts, jdbcOptions)(sparkSession)
  }
```

**Comment by gatorsmile** [comment]
LGTM except one minor comment.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 6 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: ok

**Comment by maropu** [comment]
ping

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
The build passed. The tests have passed in the previous run. The current tests will be killed at the midnight. 

LGTM

Thanks! Merged to master.

**Pull request closed without merging**